src/Lib.P2P/PeerConnection.cs

```diff
@@ -0,0 +1,433 @@
+            CancellationToken cancel = default)
+        {
+            await EstablishProtocolAsync("/multistream/", cancel).ConfigureAwait(false);
+
+            // Find the first security protocol that is also supported by the remote.
+            var exceptions = new List<Exception>();
+            foreach (var protocol in securityProtocols)
+            {
+                try
+                {
+                    await EstablishProtocolAsync(protocol.ToString(), cancel).ConfigureAwait(false);
+                }
+                catch (Exception e)
+                {
+                    exceptions.Add(e);
+                    continue;
+                }
+
+                await protocol.EncryptAsync(this, cancel).ConfigureAwait(false);
+                break;
+            }
+
+            if (!SecurityEstablished.Task.IsCompleted)
+            {
+                throw new AggregateException("Could not establish a secure connection.", exceptions);
+            }
+
+            await EstablishProtocolAsync("/multistream/", cancel).ConfigureAwait(false);
+            await EstablishProtocolAsync("/mplex/", cancel).ConfigureAwait(false);
+
+            var muxer = new Muxer
+            {
+                Channel = Stream,
+                Initiator = true,
+                Connection = this
+            };
+            muxer.SubstreamCreated += (s, e) => _ = ReadMessagesAsync(e, CancellationToken.None);
+            MuxerEstablished.SetResult(muxer);
+
+            _ = muxer.ProcessRequestsAsync(cancel);
+        }
+
+        /// <summary>
+        ///   TODO:
+        /// </summary>
+        /// <param name="name"></param>
+        /// <param name="cancel"></param>
+        /// <returns></returns>
+        internal Task EstablishProtocolAsync(string name, CancellationToken cancel)
+        {
+            return EstablishProtocolAsync(name, Stream, cancel);
+        }
+
+        /// <summary>
+        ///   TODO:
+        /// </summary>
+        /// <param name="name"></param>
+        /// <param name="stream"></param>
+        /// <param name="cancel"></param>
+        /// <returns></returns>
+        public async Task EstablishProtocolAsync(string name,
+            Stream stream,
+            CancellationToken cancel = default(CancellationToken))
+        {
+            var protocols = ProtocolRegistry.Protocols.Keys
+               .Where(k => k == name || k.StartsWith(name))
+               .Select(k => VersionedName.Parse(k))
+               .OrderByDescending(vn => vn)
+               .Select(vn => vn.ToString());
+            foreach (var protocol in protocols)
+            {
+                await Message.WriteAsync(protocol, stream, cancel).ConfigureAwait(false);
+                var result = await Message.ReadStringAsync(stream, cancel).ConfigureAwait(false);
+                if (result == protocol) return;
+            }
+
+            if (protocols.Count() == 0) throw new Exception($"Protocol '{name}' is not registered.");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)